bin/svndbadmin

```diff
@@ -78,237 +80,246 @@ def _to_str(bs):
+    return
+
+  for (path, action, plus, minus) in revision.changes:
+    directory, file = os.path.split(path)
+    commit = cvsdb.CreateCommit()
+    commit.SetRepository(repo.path)
+    commit.SetDirectory(directory)
+    commit.SetFile(file)
+    commit.SetRevision(str(rev))
+    commit.SetAuthor(revision.author)
+    commit.SetDescription(revision.log)
+    commit.SetTime(revision.date)
+    commit.SetPlusCount(plus)
+    commit.SetMinusCount(minus)
+    commit.SetBranch(None)
+
+    if action == 'add':
+      commit.SetTypeAdd()
+    elif action == 'remove':
+      commit.SetTypeRemove()
+    elif action == 'change':
+      commit.SetTypeChange()
+
+    if command == 'update':
+      result = db.CheckCommit(commit)
+      if result and not force:
+        continue  # already recorded
+
+    # commit to database
+    db.AddCommit(commit)
+    committed = 1
+
+  if verbose:
+    if committed:
+      _print("done.")
+    else:
+      _print("skipped (already recorded).")
+
 
 def main(command, repository, revs=[], verbose=0, force=0):
-    cfg = viewvc.load_config(CONF_PATHNAME)
-    db = cvsdb.ConnectDatabase(cfg)
-
-    # Purge what must be purged.
-    if command in ('rebuild', 'purge'):
-        if verbose:
-            _print(f"Purging commit info for repository root `{repository}'")
-        try:
-            db.PurgeRepository(repository)
-        except cvsdb.UnknownRepositoryError as e:
-            if command == 'purge':
-                sys.stderr.write("ERROR: " + str(e) + "\n")
-                sys.exit(1)
-
-    # Record what must be recorded.
-    if command in ('rebuild', 'update'):
-        if not os.path.exists(repository):
-            sys.stderr.write(f"ERROR: could not find repository {repository}\n")
-            sys.exit(1)
-        repo = SvnRepo(repository)
-        if command == 'rebuild' or (command == 'update' and not revs):
-            for rev in range(repo.rev_max+1):
-                handle_revision(db, command, repo, rev, verbose)
-        elif command == 'update':
-            if revs[0] is None:
-                revs[0] = repo.rev_max
-            if revs[1] is None:
-                revs[1] = repo.rev_max
-            revs.sort()
-            for rev in range(revs[0], revs[1]+1):
-                handle_revision(db, command, repo, rev, verbose, force)
+  cfg = viewvc.load_config(CONF_PATHNAME)
+  db = cvsdb.ConnectDatabase(cfg)
+
+  # Purge what must be purged.
+  if command in ('rebuild', 'purge'):
+    if verbose:
+      _print(f"Purging commit info for repository root `{repository}'")
+    try:
+      db.PurgeRepository(repository)
+    except cvsdb.UnknownRepositoryError as e:
+      if command == 'purge':
+        sys.stderr.write("ERROR: " + str(e) + "\n")
+        sys.exit(1)
+
+  # Record what must be recorded.
+  if command in ('rebuild', 'update'):
+    if not os.path.exists(repository):
+      sys.stderr.write(
+        f"ERROR: could not find repository {repository}\n")
```

Comment:
Though it is off topic, f-strings was introduced in Python 3.6, but our INSTALL file says ViewVC supports Python 3.3+.  However I think it is reasonable to drop support for Python 3.5 and below because those Pythons are already in EOL.

Reply:
I bumped the minimum Python to 3.6 on `master`.